### DISCUSSION TITLE
In clickhouse, is there a requirement for the length of the primary key for performance reasons?

### DISCUSSION THREAD

**Discussion by wymli** [ClickHouse/ClickHouse#51763] [Q&A] [answered]
For example, I have a primary key of type string, about `60 bytes`. 
Will this affect the storage space or search speed? Should I shorten it?
BTW, I want to add an index to the starting position of each primary key, how should I do it? I want to be able to quickly locate the starting position of the primary key.

**Comment by DanRoscigno** [comment, marked as answer]
@wymli I would recommend reading this doc, there are several good examples. https://clickhouse.com/docs/en/optimize/sparse-primary-indexes

**Comment by den-crane** [comment]
Long Primary index can drastically reduce performance https://github.com/ClickHouse/ClickHouse/issues/6924
I suggest to keep primary index to 4-5 columns with small datatypes as possible (avoid long strings).
Primary index is stored in memory and long primary index consumes RAM excessively.
Also keep in mind that `ORDER BY` <> `PRIMARY KEY` , so you may have `MergeTree ORDER BY (a,b,c,d,e,f,g,k,l,n,m,w,y) PRIMARY KEY (a,b,c,d)`

https://clickhousedb.slack.com/archives/CU478UEQZ/p1686594136721999?thread_ts=1686568501.389679&cid=CU478UEQZ
